exdir/core/dataset.py

```diff
@@ -125,13 +123,15 @@ def _reload_data(self):
                 else:
                     raise e
 
+    @assert_file_open
     def _reset_data(self, value, attrs, meta):
         self._data_memmap = np.lib.format.open_memmap(
             self.data_filename,
             mode="w+",
             dtype=value.dtype,
             shape=value.shape
         )
+        self.file._open_datasets[self.name] = self._data_memmap
```

Comment:
This might be an issue: We will now hold a reference to all datasets that have ever been opened, even if they are not in use anywhere in the code. This means that the user now _must_ close the file to free the datasets from memory, while previously it was enough that the datasets went out of scope and the GC cleaned it up. I think file._open_datasets should use a weak reference here (probably a WeakValueDictionary): https://docs.python.org/3/library/weakref.html

Also, it would be better if this pointed to self rather than the internal self._data_memmap. Then you only need to do this once and not update the reference on every reset/reload.